src/ir/ir_traversing_visitor.h

```diff
@@ -24,47 +25,74 @@ namespace raksha::ir {
 // A visitor that also traverses the children of a node and allows performing
 // different actions before (PreVisit) and after (PostVisit) the children are
 // visited. Override any of the `PreVisit` and `PostVisit` methods as needed.
-template <typename Derived>
-class IRTraversingVisitor : public IRVisitor<Derived> {
+template <typename Derived, typename Result = Unit>
+class IRTraversingVisitor : public IRVisitor<Derived, Result> {
  public:
   virtual ~IRTraversingVisitor() {}
 
   // Invoked before all the children of `module` is visited.
-  virtual void PreVisit(const Module& module) {}
+  virtual Result PreVisit(const Module& module) {
+    return Result();
+  }
   // Invoked after all the children of `module` is visited.
-  virtual void PostVisit(const Module& module) {}
+  virtual Result PostVisit(const Module& module, Result result) {
+    return result;
+  }
+  virtual Result FoldResult(const Module& module, Result child_result,
+                            Result result) {
+    return result;
+  }
   // Invoked before all the children of `block` is visited.
-  virtual void PreVisit(const Block& block) {}
+  virtual Result PreVisit(const Block& block) {
+    return Result();
+  }
   // Invoked after all the children of `block` is visited.
-  virtual void PostVisit(const Block& block) {}
+  virtual Result PostVisit(const Block& block, Result result) {
+    return result;
+  }
+  virtual Result FoldResult(const Block& block, Result child_result,
+                            Result result) {
+    return result;
+  }
   // Invoked before all the children of `operation` is visited.
-  virtual void PreVisit(const Operation& operation) {}
+  virtual Result PreVisit(const Operation& operation) {
+    return Result();
+  }
   // Invoked after all the children of `operation` is visited.
-  virtual void PostVisit(const Operation& operation) {}
+  virtual Result PostVisit(const Operation& operation, Result result) {
+    return result;
+  }
+  virtual Result FoldResult(const Operation& operation, Result child_result,
+                            Result result) {
+    return result;
+  }
 
-  void Visit(const Module& module) final override {
-    PreVisit(module);
+  Result Visit(const Module& module) final override {
+    Result result = PreVisit(module);
     for (const std::unique_ptr<Block>& block : module.blocks()) {
-      block->Accept(*this);
+      result =
+          FoldResult(*block, block->Accept<Derived, Result>(*this), std::move(result));
     }
-    PostVisit(module);
+    return PostVisit(module, std::move(result));
```

Comment:
Why not use `fold` from our `utils` package for this and other folds in the visitor? You could write the whole function like so:

```C++
Result pre_visit_result = PreVisit(module);
Result fold_result = fold(module.blocks(), std::move(pre_visit_result),
       [&this](Result acc, const std::unique_ptr<Block> &block) {
         return FoldResult(*block, block->Accept<Derived, Result>(*this), std::move(acc));
       });
return PostVisit(module, std::move(fold_result));
```

This gets rid of the loop and the mutable `result` local. Each of these locals, of course, could be inlined to get rid of the `std::move`, but this is closest to your current form (and possibly the most readable form).

Reply:
One... 'very good reason'

...I didn't know about it! :)

Thanks great. I'll change it right away